qiime2/sdk/usage.py

```diff
@@ -0,0 +1,368 @@
+            opts[self.get(output)] = output
+
+        return opts
+
+
+class ScopeRecord:
+    STATE_PENDING = 'pending'
+    STATE_FINISHED = 'finished'
+
+    def __init__(self, id: str, factory: callable = None,
+                 value: object = None,
+                 assert_has_line_matching: callable = None):
+        if factory is not None and not callable(factory):
+            # TODO
+            raise Exception
+
+        if assert_has_line_matching is not None and \
+                not callable(assert_has_line_matching):
+            # TODO
+            raise Exception
+
+        self.id = id
+        if value is None:
+            self.state = self.STATE_PENDING
+        else:
+            self.state = self.STATE_FINISHED
+        self._factory = factory
+        self._result = value
+        self._assert_has_line_matching_ = assert_has_line_matching
+
+    def __repr__(self):
+        if self.state == self.STATE_PENDING:
+            return 'ScopeRecord<id=%s, state=%s>' % (self.id, self.state)
+        else:
+            return 'ScopeRecord<id=%s, result=%r>' % (self.id, self.result)
+
+    @property
+    def result(self):
+        if self.is_pending:
+            raise Exception
+        return self._result
+
+    @property
+    def is_pending(self):
+        return self.state == self.STATE_PENDING
+
+    def resolve(self):
+        if self.is_pending:
+            self._result = self._factory()
+            self.state = self.STATE_FINISHED
+
+    def assert_has_line_matching(self, label, path, expression):
+        return self._assert_has_line_matching_(self.id, label, path,
+                                               expression)
+
+
+class Scope:
+    def __init__(self):
+        self._records = dict()
+
+    def __repr__(self):
+        return '%r' % (self._records, )
+
+    @property
+    def records(self):
+        return types.MappingProxyType(self._records)
+
+    def push_record(self, ref, factory=None, value=None,
+                    assert_has_line_matching=None):
+        record = ScopeRecord(id=ref, factory=factory, value=value,
+                             assert_has_line_matching=assert_has_line_matching)
+        self._records[ref] = record
+        return record
+
+    def get_record(self, id):
+        try:
+            return self.records[id]
+        except KeyError:
+            raise KeyError('No record with id: "%s" in scope.' % (id,))
+
+
+class Usage(metaclass=abc.ABCMeta):
+    def __init__(self):
+        self._scope = Scope()
+
+    def init_data(self, ref, factory):
+        override = self._factory_override(ref, factory)
+        return self._push_record(ref=ref, factory=override)
+
+    def merge_metadata(self, ref, *records):
+        if len(records) < 2:
+            # TODO: seems like we should error for this case - thoughts?
+            raise ValueError('')
+
+        factory = self._merge_metadata_(ref, records)
+        return self.init_data(ref, factory)
```

Comment:
Is there any risk of the `_override_factory` trampling on the intention of `_merge_metadata_`? Since we are returning a closure, I don't think I would anticipate it running through another factory method.

Reply:
I'm imagining a situation where `_merge_metadata_` returns a closure that has a tuple of references, but my `_override_factory` drops the factory and uses `ref`. So now I  have to figure out some kind of type I can check for inside the `_override_factory` to avoid discarding it. It seems conceptually easier if `_merge_metadata_` had the same rules as `_override_factory` (and maybe we call that `_init_data_` like I mentioned elsewhere).

Reply:
I think this makes sense, because they are both really a de-reference operation (one is just more involved).

Reply:
> It seems conceptually easier if `_merge_metadata_`

I actually had that laying around in my working changeset initially, then changed my mind. I think the change you suggested sounds good, and works really well with the `_init_data_` change above.